app/src/main/java/com/github/polybooks/core/database/implementation/OLBookDatabase.kt

```diff
@@ -80,30 +82,25 @@ class OLBookDatabase(private val url2json : (String) -> CompletableFuture<JsonEl
         override fun getAll(): CompletableFuture<List<Book>> {
             if (empty) return CompletableFuture.completedFuture(Collections.emptyList())
             else {
-                assert(isbn != null)
-                val url = isbn2URL(isbn!!)
-                return url2json(url)
-                        .thenApply { parseBook(it) }
-                        .thenCompose { updateBookWithAuthorName(it) }
-                        .thenApply { listOf(it) }
-                        .exceptionally { exception ->
-                            if (exception is CompletionException && exception.cause is FileNotFoundException)
-                                return@exceptionally Collections.emptyList<Book>()
-                            else if (exception is CompletionException) throw exception.cause!!
-                            else throw exception
-                        }
+                assert(isbns != null)
+                val futures = isbns!!.map{getBookByISBN(it)}
+                return listOfFuture2FutureOfList(futures).thenApply { it.filterNotNull() }
             }
         }
 
         @RequiresApi(Build.VERSION_CODES.N)
         override fun getN(n: Int, page: Int): CompletableFuture<List<Book>> {
-            if (n < 0 || page < 0) {
+            if (n <= 0 || page < 0) {
                 throw IllegalArgumentException(
-                    if (n < 0) "Cannot return a negative ($n) number of results"
+                    if (n <= 0) "Cannot return a negative/null ($n) number of results"
                     else "Cannot return a negative ($page) page number"
                 )
             }
-            return getAll()
+            return getAll().thenApply { list ->
+                val lowRange = min(n*page, list.size)
+                val highRange = min(n*page + n, list.size)
+                list.subList(lowRange, highRange)
+            }
```

Comment:
Might have an issue if the list has size 0. In this case, then statement `list.subList(0, 0)` will always be executed. The first index is inclusive while the second is exclusive

Reply:
I think that since the second is exclusive, it will return an empty list, which is what we want. IIRC, there are tests that get empty lists, so that case should be tested.